File: examples/01-singleton/after.ts
/**
 * Video 01 — Singleton, with the pattern.
 *
 * The diff against before.ts is four lines: one import, one `singleton(...)`, and
 * the call sites losing three characters each. `loadConfig` itself is untouched, so
 * it stays testable on its own.
 *
 * Run it:  node examples/01-singleton/after.ts
 */
import assert from 'node:assert/strict';
import { singleton } from '../../packages/ts/src/index.ts';

let fileOnDisk = '{"currency":"MXN","retries":3}';
let reads = 0;

type Config = { currency: string; retries: number };

function loadConfig(): Config {
  reads += 1;
  return JSON.parse(fileOnDisk) as Config;
}

// The whole change. `loadConfig` decides *what* to load; `singleton` decides *when*.
const config = singleton(loadConfig);

function handleCheckout(): string {
  return `charging in ${config().currency}`;
}

function handleRefund(): string {
  return `refunding in ${config().currency}`;
}

function logStartup(): string {
  return `retries = ${config().retries}`;
}

const checkout = handleCheckout();
fileOnDisk = '{"currency":"USD","retries":3}'; // the file still changes...
const refund = handleRefund();
logStartup();

console.log(`config leído del disco: ${reads} vez`);
console.log(`  checkout: ${checkout}`);
console.log(`  refund:   ${refund}`);
console.log(`  las tres llamadas ven el mismo objeto: ${config() === config()}`);

assert.equal(reads, 1, 'read once, no matter how many callers');
assert.equal(checkout, 'charging in MXN');
assert.equal(refund, 'refunding in MXN'); // ...and now they agree
assert.equal(config(), config(), 'the same object, so they cannot disagree');

// The part that saves your test suite. A hand-rolled singleton in a module-level
// variable leaks state from one test into the next; this drops it.
config.reset();
const afterReset = config();
assert.equal(reads, 2, 'reset means the next call builds again');
assert.equal(afterReset.currency, 'USD', 'and it picks up the current file');
console.log('\n✅ una lectura, una sola verdad, y reset() para los tests');

File: examples/01-singleton/after.py
"""Video 01 — Singleton, with the pattern.

The diff against before.py is two lines: one import and one `singleton(...)`. The
call sites lose a pair of parentheses. `load_config` itself is untouched.

Run it:  PYTHONPATH=packages/py python3 examples/01-singleton/after.py
"""

import functools
import json

from gof_patterns import singleton

file_on_disk = '{"currency":"MXN","retries":3}'
reads = 0


def load_config() -> dict:
    global reads
    reads += 1
    return json.loads(file_on_disk)


# The whole change. `load_config` decides *what* to load; `singleton` decides *when*.
config = singleton(load_config)


def handle_checkout() -> str:
    return f"charging in {config()['currency']}"


def handle_refund() -> str:
    return f"refunding in {config()['currency']}"


def log_startup() -> str:
    return f"retries = {config()['retries']}"


checkout = handle_checkout()
file_on_disk = '{"currency":"USD","retries":3}'  # the file still changes...
refund = handle_refund()
log_startup()

print(f"config leído del disco: {reads} vez")
print(f"  checkout: {checkout}")
print(f"  refund:   {refund}")
print(f"  las tres llamadas ven el mismo objeto: {config() is config()}")

assert reads == 1, "read once, no matter how many callers"
assert checkout == "charging in MXN"
assert refund == "refunding in MXN"  # ...and now they agree
assert config() is config(), "the same object, so they cannot disagree"

# The part that saves your test suite.
config.reset()
after_reset = config()
assert reads == 2, "reset means the next call builds again"
assert after_reset["currency"] == "USD", "and it picks up the current file"

# What the README tells you to prefer when the function is yours to decorate.
# Same pattern, no dependency — and this is the honest version of the video.
reads_stdlib = 0


@functools.cache
def config_stdlib() -> dict:
    global reads_stdlib
    reads_stdlib += 1
    return json.loads(file_on_disk)


assert config_stdlib() is config_stdlib()
assert reads_stdlib == 1
config_stdlib.cache_clear()  # this is `reset`
config_stdlib()
assert reads_stdlib == 2

print("\n✅ una lectura, una sola verdad, y reset() para los tests")
print("   (y con functools.cache es lo mismo si la función es tuya)")
